webapp/src/ts/services/db-sync.service.ts

```diff
@@ -117,58 +99,118 @@ export class DBSyncService {
 
+    const options = {
+      filter: readOnlyFilter,
+      batch_size: batchSize
+    };
+
     const remote = this.dbService.get({ remote: true });
-    const options = Object.assign({}, direction.options, { batch_size: batchSize });
     return this.dbService.get()
-      .replicate[direction.name](remote, options)
-      .on('change', replicationResult => {
-        if (direction.onChange) {
-          direction.onChange(replicationResult);
-        }
-      })
+      .replicate
+      .to(remote, options)
       .on('denied', (err) => {
-        console.error(`Denied replicating ${direction.name} remote server`, err);
-        if (direction.onDenied) {
-          direction.onDenied(err);
-        }
+        console.error(`Denied replicating to remote server`, err);
+        this.dbSyncRetryService.retryForbiddenFailure(err);
         telemetryEntry.recordDenied();
       })
       .on('error', (err) => {
-        console.error(`Error replicating ${direction.name} remote server`, err);
+        console.error(`Error replicating to remote server`, err);
         telemetryEntry.recordFailure(err, this.knownOnlineState);
       })
       .then(info => {
-        console.debug(`Replication ${direction.name} successful`, info);
+        console.debug(`Replication to successful`, info);
         telemetryEntry.recordSuccess(info);
       })
       .catch(err => {
-        if (err.code === 413 && direction.name === 'to' && batchSize > 1) {
+        if (err.code === 413 && batchSize > 1) {
           batchSize = Math.floor(batchSize / 2);
           console.warn('Error attempting to replicate too much data to the server. ' +
             `Trying again with batch size of ${batchSize}`);
-          return this.replicate(direction, { batchSize });
+          return this.replicateToRetry({ batchSize });
         }
-        console.error(`Error replicating ${direction.name} remote server`, err);
-        return direction.name;
+        console.error(`Error replicating to remote server`, err);
+        throw err;
       });
   }
 
-  private replicateIfAllowed(direction) {
-    return direction.allowed().then(allowed => {
-      if (!allowed) {
-        // not authorized to replicate - that's ok, skip silently
-        return;
-      }
-      return this.replicate(direction);
-    });
+  private async replicateTo() {
+    if (!await this.authService.has('can_edit')) {
+      // not authorized to replicate - that's ok, skip silently
+      return;
+    }
+
+    try {
+      await this.replicateToRetry();
+    } catch (err) {
+      return err;
+    }
+  }
+
+  private async downloadDocsBatch (batch) {
+    if (!batch.length) {
+      return;
+    }
+
+    const res = await this.dbService.get({ remote: true }).bulkGet({ docs: batch, attachments: true });
+    const docs = res.results
+      .map(result => result.docs && result.docs[0] && result.docs[0].ok)
+      .filter(doc => doc);
+    await this.dbService.get().bulkDocs(docs, { new_edits: false });
```

Comment:
I still have concerns about partial downloads, for example, if the user has 200 docs to download, gets the first batch, then disconnects. This will leave the db in an inconsistent state. The chances of having to download > 100 is probably quite low, and it was possible to happen even with the seq ordered approach but I think we've increased the chance with this approach.

We could...

1. sort the doc ids so we get system docs, then contacts, then reports, then tasks, but that's crude at best
2. store the docs in a temporary location (maybe memory but an on disk db would probably be required) and only copy them over once all batches complete, but that's complicated.
3. just roll with it and it'll be eventually consistent (the approach right now).

What do you think? Can you convince me I'm over-engineering it?

Reply:
I believe the argument I used in Nairobi to convince you that you're over engineering it with transactional approach is that the changes feed, if we were to follow vanilla replication protocol, is not ordered. It will pick an amount of docs from each shard, starting from the seq provided, and then return the new seq. The amount of docs picked from each shard every time we request is not deterministic, and you will get different results every time you request a low number of changes starting from a specific seq. 
Vanilla replication is eventually consistent, and we were already in this situation. 

Additionally, the order in which the changes feed is sorted is not necessarily consistent with CHT workflows, because you can have a new contact, a new report for this contact, and then have the contact edited later on. The report will be received much earlier because the contact was pushed to the end of the changes feed. 
So another argument for: we were already inconsistent and relying on eventual consistency. 

Convinced?

Reply:
Yeah I think so. That and the fact that most users won't exceed the BATCH_SIZE means they'll get everything in one go anyway. I still think we've made it worse, but not by enough. Let's go with it as is.